src/tree/ResourceCache.ts

```diff
@@ -0,0 +1,120 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { ResourceGroup } from "@azure/arm-resources";
+import { IActionContext, nonNullProp } from "@microsoft/vscode-azext-utils";
+import { AppResource } from "@microsoft/vscode-azext-utils/hostapi";
+import { ThemeIcon } from "vscode";
+import { GroupBySettings } from "../commands/explorer/groupBy";
+import { createAzureExtensionsGroupConfig } from "../utils/azureUtils";
+import { localize } from "../utils/localize";
+import { settingUtils } from "../utils/settingUtils";
+import { AppResourceTreeItem } from "./AppResourceTreeItem";
+import { GroupTreeItemBase } from "./GroupTreeItemBase";
+import { ResourceGroupTreeItem } from "./ResourceGroupTreeItem";
+import { SubscriptionTreeItem } from "./SubscriptionTreeItem";
+
+export type GroupTreeMap = { [groupTreeId: string]: GroupTreeItemBase };
+export class ResourceCache {
+    private _subscriptionTreeItem: SubscriptionTreeItem;
+    private _resourceGroups: ResourceGroup[];
+    private _appResources: AppResourceTreeItem[];
+
+    public constructor(sub: SubscriptionTreeItem) {
+        this._subscriptionTreeItem = sub;
+        this.resetCache();
+    }
+
+    private resetCache(): void {
+        this._appResources = [];
+        this._resourceGroups = [];
+    }
+
+    public get appResources(): AppResourceTreeItem[] {
+        return this._appResources;
+    }
+
+    public set appResources(resources: AppResource[]) {
+        const addedResources = this.resolveProvidedResources(this._appResources, resources);
+        this._appResources.push(...addedResources.map(added => AppResourceTreeItem.Create(this._subscriptionTreeItem, added)));
+    }
+
+    public set resourceGroups(resourceGroups: ResourceGroup[]) {
+        this._resourceGroups.push(...this.resolveProvidedResources(this._resourceGroups, resourceGroups));
+    }
+
+    public getTreeMap(context: IActionContext): GroupTreeMap {
+        const treeMap: GroupTreeMap = {};
+        // start this as early as possible
+        const getResourceGroupTask: (resourceGroup: string) => Promise<ResourceGroup | undefined> = async (resourceGroup: string) => {
```

Comment:
I don't think this will start it, just create the promise (referring to the comment just above).